### PR TITLE
GH-37367: [MATLAB] Add `arrow.array.Date32Array` class

### LINKED ISSUE #37367: [MATLAB] Add `arrow.array.Date32Array` class
### Describe the usage question you have. Please include as many useful details as  possible.

Now that `arrow.type.Date32Type` class has been added to the MATLAB Interface (#37348), we can add the `arrow.array.Date32Array` class.

### Component(s)

MATLAB

### PR DESCRIPTION
**Pull request by kevingurney**
### Rationale for this change

Now that `arrow.type.Date32Type` class has been added to the MATLAB Interface (#37348), we can add the `arrow.array.Date32Array` class.

`Date32Array`s can be created from MATLAB [`datetime`](https://www.mathworks.com/help/matlab/ref/datetime.html) values. 

### What changes are included in this PR?

1. Added a new `arrow.array.Date32Array` class.
2. Added a new `arrow.type.traits.Date32Traits` class.
3. Added `arrow.type.Date32Type` support to `arrow.type.traits.traits` function.
4. Fixed typo `arrray` in `tTime32Array` test class.
5. Fixed bug in `numeric_array.h` where the `CType` rather than the `ArrowType` was being used to determine the `DataType` of an array class that is a `NumericArray<T>`.

`Date32Array`s can be created from MATLAB [`datetime`](https://www.mathworks.com/help/matlab/ref/datetime.html) values using the `fromMATLAB` method. `Date32Array`s can be converted to MATLAB `datetime` values using the `toMATLAB` method.

**Example**
```matlab
>> dates = [datetime(2021, 1, 2, 3, 4, 5), datetime(2023, 1, 1), datetime(1989, 2, 3, 10, 10, 10)]'

dates = 

  3x1 datetime array

   02-Jan-2021 03:04:05
   01-Jan-2023 00:00:00
   03-Feb-1989 10:10:10

>> array = arrow.array.Date32Array.fromMATLAB(dates)                                               

array = 

[
  2021-01-02,
  2023-01-01,
  1989-02-03
]

>> array.toMATLAB()                                                                                

ans = 

  3x1 datetime array

   02-Jan-2021
   01-Jan-2023
   03-Feb-1989
``` 

### Are these changes tested?

Yes.

1. Added a new `tDate32Array` test class.
2. Added `Date32` related test to `ttraits.m`.
6. Added a new `tDate32Traits.m` test class.

### Are there any user-facing changes?

Yes.

1. Users can now create `arrow.array.Date32Array`s  from MATLAB `datetime`s.

### Future Directions

1. #37230
2. Add `arrow.array.Date64Array`.
3. Add a way to extract the raw `int32` values from an `arrow.array.Date32Array` without converting to a MATLAB `datetime` using `toMATLAB`.

### Notes

1. Thank you @sgilmore10 for your help with this pull request!
* Closes: #37367

### CODE REVIEW

File: matlab/src/cpp/arrow/matlab/array/proxy/numeric_array.h [hunk 1]
```diff
@@ -56,7 +56,7 @@ class NumericArray : public arrow::matlab::array::proxy::Array {
             
             auto data_buffer = std::make_shared<MatlabBuffer>(numeric_mda);
 
-            const auto data_type = arrow::CTypeTraits<CType>::type_singleton();
+            const auto data_type = arrow::TypeTraits<ArrowType>::type_singleton();
```
Comment (sgilmore10) [inline matlab/src/cpp/arrow/matlab/array/proxy/numeric_array.h:59]: Thanks for catching this!

File: matlab/src/matlab/+arrow/+array/Date32Array.m [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+            validElements = arrow.internal.validate.parseValidElements(data, opts);
+
+            % UNIX Epoch (January 1st, 1970)
+            if ~isempty(data.TimeZone)
+                unixEpoch = datetime(0, ConvertFrom="posixtime", TimeZone="UTC");
+            else
+                unixEpoch = datetime(0, ConvertFrom="posixtime");
+            end
+
+            numDays = int32(floor(days(data - unixEpoch)));
```
Comment (sgilmore10) [inline matlab/src/matlab/+arrow/+array/Date32Array.m:76]: Might want to leave a comment as to why we are `floor`ing the result of `data - unixEpoch`.

**Review by sgilmore10 — APPROVED** [review]
Looks good!

File: matlab/src/matlab/+arrow/+array/Date32Array.m (hunk 2 above)
Comment (kevingurney) [reply matlab/src/matlab/+arrow/+array/Date32Array.m:84]: Fixed.

**Review by kou — APPROVED** [review]
+1

File: matlab/src/matlab/+arrow/+array/Date32Array.m [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+            unixEpoch = datetime(0, ConvertFrom="posixtime");
+            % Date32 value represents the number of days before
+            % or after the Unix Epoch. This works for negative values
+            % too.
+            dates = unixEpoch + days(matlabArray);
+            dates(~obj.Valid) = obj.NullSubstitutionValue;
+        end
+
+        function times = datetime(obj)
+            times = obj.toMATLAB();
```
[the remark this answers is not in the document]
Comment (kevingurney) [reply matlab/src/matlab/+arrow/+array/Date32Array.m:47]: Thanks for catching this! I meant to name this `dates`.

[the remark this answers is not in the document]
Comment (kevingurney) [reply matlab/src/matlab/+arrow/+array/Date32Array.m]: Fixed,

**Pull request merged**